recap/src/Imp/Eval.hs

```diff
@@ -1,30 +1,74 @@
 module Imp.Eval where
 
 import Control.Monad.State
-    ( modify, evalStateT, MonadState(get), MonadTrans(lift), StateT )
-import Control.Monad.Trans.Except ( runExceptT, throwE, ExceptT )
+  ( MonadState (get),
+    MonadTrans (lift),
+    StateT,
+    evalStateT,
+    modify,
+  )
+import Control.Monad.Trans.Except (ExceptT, runExceptT, throwE)
 import qualified Data.Map.Strict as M
+import Imp.Lang
 import Text.Printf (printf)
 import Text.Read (readMaybe)
-import Imp.Lang
 
 -- Variable mapping; shadowing is allowed
 type VarMap = M.Map String Int
 
--- Possible errors reported during evaluation 
+-- Possible errors reported during evaluation
 data Error = UndefinedVar String | ParsingErr String | DivByZero deriving (Show)
 
 -- Evaluation monad
 type EvalM = StateT VarMap (ExceptT Error IO)
 
 evalExpr :: Expr -> EvalM Int
-evalExpr = undefined 
+evalExpr (Var v) = do
+  m <- get
+  case m M.!? v of
+    Nothing -> lift $ throwE $ UndefinedVar v
+    Just x -> return x
+evalExpr (Const x) = return x
+evalExpr (BinOp op l r) = do
+  x <- evalExpr l
+  y <- evalExpr r
+  x `opF` y
+  where
+    opF = case op of
+      Plus -> (return .) . (+)
+      Minus -> (return .) . (-)
+      Mult -> (return .) . (*)
+      Div -> \x y ->
+        if y /= 0
+          then
+            return $ x `div` y
+          else
+            lift $ throwE DivByZero
```

Comment:
This is not a criticism, just a comment: there is a missed opportunity to short-circuit here. There is no need to evaluate `x`, if `y` is `0`. This would mean loss of generality, of course, but...